### PR TITLE
Document open_timeout and read_timeout for SMTP Settings on ActionMailer 

### LINKED ISSUE #41244: ActionMailer does not timeout network connections by default
I have a Sidekiq customer complaining of a lot of "stuck" threads. He sends a lot of email so I suspected misbehaving SMTP servers.

The `mail` gem has support for timeouts here but they default to `nil`:

https://github.com/mikel/mail/blob/7b3e100f42f2d7738c3af7bf1909777568270b67/lib/mail/network/delivery_methods/smtp.rb#L91

Rails does not appear to set them by default:

https://github.com/rails/rails/blob/a0908477cfe2c52ffbb1a0a18a064b6dd7c255a2/actionmailer/lib/action_mailer/delivery_methods.rb#L20-L27

nor does it appear to document how to set timeouts in the SMTP configuration section:

https://guides.rubyonrails.org/action_mailer_basics.html#action-mailer-configuration

Would a read/open timeout of 5 seconds be reasonable to use as a default going forward? Could we update the ActionMailer guide for 6.1 so people know how to configure the timeouts?

### PR DESCRIPTION
**Pull request by andrehjr**
### Summary

Add documentation for open_timeout/read_timeout on ActionMailer. 

Closes #41244

### CODE REVIEW

**Review by rafaelfranca — COMMENTED** [review]
I think we should only change those default when the application is using the Rails 6.2 defaults. This means that we should not change `delivery_methods.rb` but instead change `config.action_mailer.smtp_settings` in the railtie, only when the default is for 6.2.

File: railties/lib/rails/application/configuration.rb [hunk 1]
```diff
@@ -210,6 +210,10 @@ def load_defaults(target_version)
             active_support.hash_digest_class = OpenSSL::Digest::SHA256
             active_support.key_generator_hash_digest_class = OpenSSL::Digest::SHA256
           end
+
+          if respond_to?(:action_mailer)
+            action_mailer.default_smtp_settings = { open_timeout: 5, read_timeout: 5 }
```
Comment (rafaelfranca) [inline railties/lib/rails/application/configuration.rb:215]: I think we should call this `default_smtp_timeout_value` and set it to be 5. If people want a different value they can use `action_mailer.smtp_settings`

File: guides/source/action_mailer_basics.md [hunk 2]
```diff
@@ -799,7 +799,7 @@ files (environment.rb, production.rb, etc...)
...
...et the password in this setting.</li><li>`:authentication` - If your mail server requires authentication, you need to specify the authentication type here. This is a symbol and one of `:plain` (will send the password in the clear), `:login` (will send password Base64 encoded) or `:cram_md5` (combines a Challenge/Response mechanism to exchange information and a cryptographic Message Digest 5 algorithm to hash important information)</li><li>`:enable_starttls_auto` - Detects if STARTTLS is enabled in your SMTP server and starts to use it. Defaults to `true`.</li><li>`:openssl_verify_mode` - When using TLS, you can set how OpenSSL checks the certificate. This is really useful if you need to validate a self-signed and/or a wildcard certificate. You can use the name of an OpenSSL verify constant ('none' or 'peer') or directly the constant (`OpenSSL::SSL::VERIFY_NONE` or `OpenSSL::SSL::VERIFY_PEER`).</li><li>`:ssl/:tls` - Enables the SMTP connection to use SMTP/TLS (SMTPS: SMTP over direct TLS connection)</li><li>`:open_timeout` - Number of seconds to wait while attempting to open a connection.</li><li>`:read_timeout` - Number of seconds to wait until timing-out a read(2) call.</li></ul>|
```
Comment (rafaelfranca) [inline guides/source/action_mailer_basics.md:802]: Just saw that in this guide has some examples of this config being set. Let's add the timeouts values there so if people copy and paste the get the right thing.

**Comment by andrehjr** [comment]
Thanks for the review @rafaelfranca! You're right, it's best to set this only as a default for 6.2.

My original change wouldn't work either (pushed too soon 😅 ), because when someone does `config.action_mailer.smtp_settings=` it ends up overriding the `smtp_settings` object.  The options defined here https://github.com/rails/rails/blob/main/actionmailer/lib/action_mailer/delivery_methods.rb#L20-L27 are dropped. 

Example:
```
#  config.action_mailer.smtp_settings =   { address: 'smtp.gmail.com' }
> ActionMailer::Base.smtp_settings
=> {:address=>"smtp.gmail.com"}
```

So I wanted to have some feedback on the path forward. I can either make the options defined on `smtp_settings` act as 'defaults' and be configured whenever someone uses `ActionMailer::Base.smtp_settings = `. Which would change how smtp_settings works. But it does seem nicer to me.

Or just push additional options via config something like https://github.com/rails/rails/pull/41248/commits/f165c0f2969c16f0cdfeaa42254807b1497b697e The downside of this change is that this will only work everything gets initialized. 

WDYT? Sorry If I missed something, or if there is a better way to do it.

**Comment by andrehjr** [comment]
Ops, I just saw your response. Will make these changes :)

**Comment by jonathanhefner** [comment]
What if we shortened the name to `action_mailer.smtp_timeout`?

**Comment by andrehjr** [comment]
I like the shortened version. Made that change and I can revert/squash later when we decide.

Pushed another commit adding (open|read)_timeout to the ActionMailer section on `guides/source/configuring.md`. I also documented the `smtp_timeout` config there!

**Comment by mperham** [comment]
Thank you @andrehjr!

I will have hundreds of customers on Rails 5.x, 6.0 and 6.1 for years to come and get reports of email sending leading to "stuck" jobs all the time. I need something to help with existing apps so here's one idea, I think they can drop this in an initializer to easily enable timeouts:

```ruby
# gem "mail", ">= 2.7.0"
Mail::SMTP::DEFAULTS[:read_timeout] = 5
Mail::SMTP::DEFAULTS[:open_timeout] = 5
```

**Comment by andrehjr** [comment]
Hi @mperham !

I think this is less invasive than adding a new configuration for ActionMailer. And older rails versions could just pull that initializer if they want. So, the idea would be to add a new initializer for 'mail.rb' setting these defaults, right?

WDYT @rafaelfranca ?

**Comment by mperham** [comment]
Well, the friendliest approach IMO is to get a Mail committer to add default timeouts here and release v2.7.2:

https://github.com/mikel/mail/blob/master/lib/mail/network/delivery_methods/smtp.rb#L91-L92

Then everyone gets this fix with `bundle up mail` and we don't have to litter code in initializers.

**Comment by andrehjr** [comment]
For now, I updated this PR to only include the documentation changes, mentioning read_timeout and open_timeout on the docs. So this part should be good to go.

If we go for adding an initializer like Mike's suggestion or use the strategy of railties and its configuration I can update this PR. 🙌

**Pull request merged**